connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala

```diff
@@ -0,0 +1,162 @@
+   *   df.groupBy("department").agg(Map(
+   *     "age" -> "max",
+   *     "expense" -> "sum"
+   *   ))
+   * }}}
+   *
+   * @since 3.4.0
+   */
+  def agg(exprs: Map[String, String]): DataFrame = {
+    toDF(exprs.map { case (colName, expr) =>
+      strToExpr(expr)(df(colName).expr)
+    }.toSeq)
+  }
+
+  /**
+   * (Java-specific) Compute aggregates by specifying a map from column name to aggregate methods.
+   * The resulting `DataFrame` will also contain the grouping columns.
+   *
+   * The available aggregate methods are `avg`, `max`, `min`, `sum`, `count`.
+   * {{{
+   *   // Selects the age of the oldest employee and the aggregate expense for each department
+   *   import com.google.common.collect.ImmutableMap;
+   *   df.groupBy("department").agg(ImmutableMap.of("age", "max", "expense", "sum"));
+   * }}}
+   *
+   * @since 3.4.0
+   */
+  def agg(exprs: java.util.Map[String, String]): DataFrame = {
+    agg(exprs.asScala.toMap)
+  }
+
+  private[this] def strToExpr(expr: String): (proto.Expression => proto.Expression) = {
+    val exprToFunc: (proto.Expression => proto.Expression) = {
+      val builder = proto.Expression.newBuilder()
+
+      (inputExpr: proto.Expression) =>
+        expr.toLowerCase(Locale.ROOT) match {
+          // We special handle a few cases that have alias that are not in function registry.
+          case "avg" | "average" | "mean" =>
+            builder.getUnresolvedFunctionBuilder
+              .setFunctionName("avg")
+              .addArguments(inputExpr)
+              .setIsDistinct(false)
+          case "stddev" | "std" =>
+            builder.getUnresolvedFunctionBuilder
+              .setFunctionName("stddev")
+              .addArguments(inputExpr)
+              .setIsDistinct(false)
+          // Also special handle count because we need to take care count(*).
+          case "count" | "size" =>
+            // Turn count(*) into count(1)
+            inputExpr match {
+              case s if s.hasUnresolvedStar =>
+                val exprBuilder = proto.Expression.newBuilder
+                exprBuilder.getLiteralBuilder.setInteger(1)
+                builder.getUnresolvedFunctionBuilder
+                  .setFunctionName("count")
+                  .addArguments(exprBuilder)
+                  .setIsDistinct(false)
+              case _ =>
+                builder.getUnresolvedFunctionBuilder
+                  .setFunctionName("count")
+                  .addArguments(inputExpr)
+                  .setIsDistinct(false)
+            }
+          case name =>
+            builder.getUnresolvedFunctionBuilder
+              .setFunctionName(name)
+              .addArguments(inputExpr)
+              .setIsDistinct(false)
+        }
+        builder.build()
+    }
+    (inputExpr: proto.Expression) => exprToFunc(inputExpr)
+  }
+}
+
+private[sql] object RelationalGroupedDataset {
+  def apply(df: DataFrame, groupingExprs: Seq[proto.Expression]): RelationalGroupedDataset = {
```

Comment:
Do we need this?

Reply:
I think without this, I will need to `new class()` to invoke the class constructor. But existing codebase does not do that but offering the object with an apply implementation. So I am following that.

Reply:
Yeah, you need to use `new`. Given that this method is not public we don't need to implement it.

Reply:
Removed this one and just use `new`.